modules/images/image-loading-optimization/storage/data.php

```diff
@@ -0,0 +1,311 @@
+ * @see ilo_verify_page_metrics_storage_nonce()
+ *
+ * @param string $slug Page metrics slug.
+ * @return string Nonce.
+ */
+function ilo_get_page_metrics_storage_nonce( string $slug ): string {
+	return wp_create_nonce( "store_page_metrics:{$slug}" );
+}
+
+/**
+ * Verify nonce for storing page metrics for a specific slug.
+ *
+ * @see wp_verify_nonce()
+ * @see ilo_get_page_metrics_storage_nonce()
+ *
+ * @param string $nonce Page metrics storage nonce.
+ * @param string $slug  Page metrics slug.
+ * @return int 1 if the nonce is valid and generated between 0-12 hours ago,
+ *             2 if the nonce is valid and generated between 12-24 hours ago.
+ *             0 if the nonce is invalid.
+ */
+function ilo_verify_page_metrics_storage_nonce( string $nonce, string $slug ): int {
+	return (int) wp_verify_nonce( $nonce, "store_page_metrics:{$slug}" );
+}
+
+/**
+ * Unshift a new page metric onto an array of page metrics.
+ *
+ * @param array $page_metrics          Page metrics.
+ * @param array $validated_page_metric Validated page metric. See JSON Schema defined in ilo_register_endpoint().
+ * @return array Updated page metrics.
+ */
+function ilo_unshift_page_metrics( array $page_metrics, array $validated_page_metric ): array {
+	array_unshift( $page_metrics, $validated_page_metric );
+	$breakpoints          = ilo_get_breakpoint_max_widths();
+	$sample_size          = ilo_get_page_metrics_breakpoint_sample_size();
+	$grouped_page_metrics = ilo_group_page_metrics_by_breakpoint( $page_metrics, $breakpoints );
+
+	foreach ( $grouped_page_metrics as &$breakpoint_page_metrics ) {
+		if ( count( $breakpoint_page_metrics ) > $sample_size ) {
+
+			// Sort page metrics in descending order by timestamp.
+			usort(
+				$breakpoint_page_metrics,
+				static function ( $a, $b ) {
+					if ( ! isset( $a['timestamp'] ) || ! isset( $b['timestamp'] ) ) {
+						return 0;
+					}
+					return $b['timestamp'] <=> $a['timestamp'];
+				}
+			);
+
+			$breakpoint_page_metrics = array_slice( $breakpoint_page_metrics, 0, $sample_size );
+		}
+	}
+
+	return array_merge( ...$grouped_page_metrics );
+}
+
+/**
+ * Gets the breakpoint max widths to group page metrics for various viewports.
+ *
+ * Each max with represents the maximum width (inclusive) for a given breakpoint. So if there is one number, 480, then
+ * this means there will be two viewport groupings, one for 0<=480, and another >480. If instead there were three
+ * provided breakpoints (320, 480, 576) then this means there will be four viewport groupings:
+ *
+ *  1. 0-320 (small smartphone)
+ *  2. 321-480 (normal smartphone)
+ *  3. 481-576 (phablets)
+ *  4. >576 (desktop)
+ *
+ * @return int[] Breakpoint max widths, sorted in ascending order.
+ */
+function ilo_get_breakpoint_max_widths(): array {
+
+	/**
+	 * Filters the breakpoint max widths to group page metrics for various viewports.
+	 *
+	 * @param int[] $breakpoint_max_widths Max widths for viewport breakpoints.
+	 */
```

Comment:
This doc block is now in the wrong place :)

Reply:
Oops! Fixed in caaef87